File: baboon-compiler/src/main/resources/baboon-runtime/rust/baboon_fixture.rs
use chrono::{DateTime, FixedOffset, NaiveDateTime, TimeZone, Utc};
use rand::Rng;
use rust_decimal::Decimal;
use std::collections::{BTreeMap, BTreeSet};
use uuid::Uuid;

pub struct BaboonRandom {
    rng: rand::rngs::ThreadRng,
}

impl BaboonRandom {
    pub fn new() -> Self {
        BaboonRandom {
            rng: rand::thread_rng(),
        }
    }

    pub fn next_i08(&mut self) -> i8 {
        self.rng.gen()
    }

    pub fn next_i16(&mut self) -> i16 {
        self.rng.gen()
    }

    pub fn next_i32(&mut self) -> i32 {
        self.rng.gen()
    }

    pub fn next_i64(&mut self) -> i64 {
        self.rng.gen()
    }

    pub fn next_u08(&mut self) -> u8 {
        self.rng.gen()
    }

    pub fn next_u16(&mut self) -> u16 {
        self.rng.gen()
    }

    pub fn next_u32(&mut self) -> u32 {
        self.rng.gen()
    }

    pub fn next_u64(&mut self) -> u64 {
        self.rng.gen()
    }

    pub fn next_f32(&mut self) -> f32 {
        self.rng.gen_range(-1000.0f32..1000.0f32)
    }

    pub fn next_f64(&mut self) -> f64 {
        self.rng.gen_range(-1000.0f64..1000.0f64)
    }

    pub fn next_f128(&mut self) -> Decimal {
        // Use integer values to avoid precision issues in round-trips
        let value: i64 = self.rng.gen_range(-999999..999999);
        Decimal::new(value, 2).normalize()
    }

    pub fn next_string(&mut self) -> String {
        let len = self.rng.gen_range(1..20);
        (0..len)
            .map(|_| {
                let idx = self.rng.gen_range(0..26);
                (b'a' + idx) as char
            })
            .collect()
    }

    pub fn next_bytes(&mut self) -> Vec<u8> {
        let len = self.rng.gen_range(1..20);
        (0..len).map(|_| self.rng.gen()).collect()
    }

    pub fn next_uid(&mut self) -> Uuid {
        Uuid::new_v4()
    }

    pub fn next_tsu(&mut self) -> DateTime<Utc> {
        // Generate a random timestamp with millisecond precision (matching other languages)
        let secs = self.rng.gen_range(0i64..2000000000i64);
        let millis = self.rng.gen_range(0u32..1000u32);
        let dt = NaiveDateTime::from_timestamp_opt(secs, millis * 1_000_000)
            .unwrap_or_else(|| NaiveDateTime::from_timestamp_opt(0, 0).unwrap());
        Utc.from_utc_datetime(&dt)
    }

    pub fn next_tso(&mut self) -> DateTime<FixedOffset> {
        let secs = self.rng.gen_range(0i64..2000000000i64);
        let millis = self.rng.gen_range(0u32..1000u32);
        let offset_hours = self.rng.gen_range(-12i32..=12i32);
        let offset = FixedOffset::east_opt(offset_hours * 3600).unwrap();
        let dt = NaiveDateTime::from_timestamp_opt(secs, millis * 1_000_000)
            .unwrap_or_else(|| NaiveDateTime::from_timestamp_opt(0, 0).unwrap());
        offset.from_utc_datetime(&dt)
    }

    pub fn next_bit(&mut self) -> bool {
        self.rng.gen()
    }

    pub fn next_usize(&mut self, max: usize) -> usize {
        self.rng.gen_range(0..max)
    }

    pub fn mk_list<T>(&mut self, mut gen: impl FnMut() -> T) -> Vec<T> {
        let count = self.rng.gen_range(0..5);
        (0..count).map(|_| gen()).collect()
    }

    pub fn mk_set<T: Ord>(&mut self, mut gen: impl FnMut() -> T) -> BTreeSet<T> {
        let count = self.rng.gen_range(0..5);
        (0..count).map(|_| gen()).collect()
    }

    pub fn mk_map<K: Ord, V>(&mut self, mut gen_key: impl FnMut() -> K, mut gen_val: impl FnMut() -> V) -> BTreeMap<K, V> {
        let count = self.rng.gen_range(0..5);
        (0..count).map(|_| (gen_key(), gen_val())).collect()
    }

    pub fn mk_option<T>(&mut self, mut gen: impl FnMut() -> T) -> Option<T> {
        if self.rng.gen_bool(0.5) {
            Some(gen())
        } else {
            None
        }
    }

    pub fn mk_enum<T: Clone>(&mut self, variants: &[T]) -> T {
        let idx = self.rng.gen_range(0..variants.len());
        variants[idx].clone()
    }
}

impl Default for BaboonRandom {
    fn default() -> Self {
        Self::new()
    }
}

File: baboon-compiler/src/main/resources/baboon-runtime/typescript/baboon_fixture.ts
import { BaboonDecimal, BaboonDateTimeUtc, BaboonDateTimeOffset } from "./baboon_runtime";

export class BaboonRandom {
    nextI08(): number {
        return Math.floor(Math.random() * 256) - 128;
    }

    nextI16(): number {
        return Math.floor(Math.random() * 65536) - 32768;
    }

    nextI32(): number {
        return Math.floor(Math.random() * 4294967296) - 2147483648;
    }

    nextI64(): bigint {
        const hi = BigInt(Math.floor(Math.random() * 4294967296));
        const lo = BigInt(Math.floor(Math.random() * 4294967296));
        const raw = (hi << 32n) | lo;
        return raw >= 9223372036854775808n ? raw - 18446744073709551616n : raw;
    }

    nextU08(): number {
        return Math.floor(Math.random() * 256);
    }

    nextU16(): number {
        return Math.floor(Math.random() * 65536);
    }

    nextU32(): number {
        return Math.floor(Math.random() * 4294967296);
    }

    nextU64(): bigint {
        const hi = BigInt(Math.floor(Math.random() * 4294967296));
        const lo = BigInt(Math.floor(Math.random() * 4294967296));
        return (hi << 32n) | lo;
    }

    nextF32(): number {
        const val = (Math.random() * 2000) - 1000;
        const buf = new Float32Array(1);
        buf[0] = val;
        return buf[0];
    }

    nextF64(): number {
        return (Math.random() * 2000) - 1000;
    }

    nextF128(): BaboonDecimal {
        const value = Math.floor(Math.random() * 1999998) - 999999;
        return BaboonDecimal.fromNumber(value / 100);
    }

    nextString(): string {
        const len = Math.floor(Math.random() * 19) + 1;
        let result = "";
        for (let i = 0; i < len; i++) {
            result += String.fromCharCode(97 + Math.floor(Math.random() * 26));
        }
        return result;
    }

    nextBytes(): Uint8Array {
        const len = Math.floor(Math.random() * 19) + 1;
        const bytes = new Uint8Array(len);
        for (let i = 0; i < len; i++) {
            bytes[i] = Math.floor(Math.random() * 256);
        }
        return bytes;
    }

    nextUid(): string {
        const hex = (n: number) => {
            let result = "";
            for (let i = 0; i < n; i++) {
                result += Math.floor(Math.random() * 16).toString(16);
            }
            return result;
        };
        return `${hex(8)}-${hex(4)}-4${hex(3)}-${(8 + Math.floor(Math.random() * 4)).toString(16)}${hex(3)}-${hex(12)}`;
    }

    nextTsu(): BaboonDateTimeUtc {
        const secs = Math.floor(Math.random() * 2000000000);
        const millis = Math.floor(Math.random() * 1000);
        return BaboonDateTimeUtc.fromMillis(secs * 1000 + millis);
    }

    nextTso(): BaboonDateTimeOffset {
        const secs = Math.floor(Math.random() * 2000000000);
        const millis = Math.floor(Math.random() * 1000);
        const offsetHours = Math.floor(Math.random() * 25) - 12;
        const epochMs = secs * 1000 + millis;
        return BaboonDateTimeOffset.fromDateAndOffset(new Date(epochMs), offsetHours * 3600 * 1000);
    }

    nextBit(): boolean {
        return Math.random() < 0.5;
    }

    nextUSize(max: number): number {
        return Math.floor(Math.random() * max);
    }

    mkEnum<T>(values: ReadonlyArray<T>): T {
        const idx = Math.floor(Math.random() * values.length);
        return values[idx];
    }
}
